### PR TITLE
Close the container if SweepReady node is used (Behind feature gate)

### PR DESCRIPTION
**Pull request by markfields**
[AB#1857](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/1857)

To help us and our partners flush out bugs with how datastores are referenced, this change adds an option which will close the container when an object is used after being marked as "SweepReady".  This means the Sweep Timeout has elapsed since it was marked as unreferenced.  Closing the container at this time will cause some user pain - users will be locked out of their documents (see follow-up PR for a way to ease that pain) - but it's only to be enabled in Dogfood, where users can reach out and help our teams get traction on these bugs.  Until all such bugs are fixed, turning on Sweep represents a Data Corruption liability.

Follow ups:
- [ ] Write e2e test to ensure only main container closes, see what summarizer does in this case (requires TestOverride.SweepTimeout)
- [ ] Add additional option to have summarizer close its parent container
- [x] Add "throttling" to container closure to only close at most 1x per <configurable timespan> -- #12004

### CODE REVIEW

File: packages/runtime/container-runtime/src/garbageCollection.ts
Comment (tyler-cai-microsoft) [inline packages/runtime/container-runtime/src/garbageCollection.ts:1357]: Are we not using something like ObjectDeletedByGCUsed? for error type? Or what Tony wrote.

Comment (markfields) [reply packages/runtime/container-runtime/src/garbageCollection.ts]: Still a draft - Just pushed updates here.

File: packages/utils/telemetry-utils/src/mockLogger.ts
Comment (tyler-cai-microsoft) [inline packages/utils/telemetry-utils/src/mockLogger.ts:85]: Didn't know this was here, this is cool

Comment (markfields) [reply packages/utils/telemetry-utils/src/mockLogger.ts:85]: Yeah there's more I want it to do, just adding bit by bit :)

**Comment by markfields** [comment]
@agarwal-navin - I'm going to do the e2e test in a separate PR since it's going to involve the test overrides for Sweep Timeout.  If you want me to block this PR on that capability, I'll pause this until that's in place, but otherwise I'll merge once you've signed off - thanks!  Looking forward to your feedback.

File: packages/runtime/container-runtime/src/garbageCollection.ts
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/garbageCollection.ts:1386]: use `UnreferencedState.SweepReady`

File: packages/runtime/container-runtime/src/garbageCollection.ts
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/garbageCollection.ts:93]: It's probably better to use an enum for this.

Comment (markfields) [reply packages/runtime/container-runtime/src/garbageCollection.ts:93]: I'm actually gonna switch it to checking the value for substring to be easier to use. And this fn will spit out an object like { mainContainer: true, summarizer: false, }

File: packages/runtime/container-runtime/src/garbageCollection.ts
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/garbageCollection.ts:341]: Consider calling this something like `objectUsedAfterGarbageCollectedError` or something that suggests this object is gone. My concern is that if this conveys that object is only marked but not gone, it diminishes the impact and makes it sound like its our bug. What do you think?

File: packages/runtime/container-runtime/src/garbageCollection.ts
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/garbageCollection.ts:93]: Also, what are the values going to be once you add summarizer? I think we should have 2 modes:
1. Summarizer and non-summarizer clients. Maybe the keyword to use here is "all" or "everywhere" or "allClients"?
2. Only summarizer clients - "summarizer" or "summarizerClient".

File: packages/runtime/container-runtime/src/garbageCollection.ts
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/garbageCollection.ts:1385]: Maybe add more context like this is a step in rolling out actual sweep where it won't even be possible to use this object since it would just be gone?

File: packages/runtime/container-runtime/src/garbageCollection.ts
Comment (tyler-cai-microsoft) [reply packages/runtime/container-runtime/src/garbageCollection.ts:93]: I'm curious as to why this feature flag is more complicated and what advantages it has over the original system.

File: packages/runtime/container-runtime/src/garbageCollection.ts
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/garbageCollection.ts:1388]: non-summarizer client may be better? client vs container.

File: packages/utils/telemetry-utils/src/mockLogger.ts
Comment (agarwal-navin) [inline packages/utils/telemetry-utils/src/mockLogger.ts:80]: I used a local version of this in the `gcInactiveNodes.spec.ts`. I will replace it with this one later :-).

File: packages/utils/telemetry-utils/src/mockLogger.ts
Comment (agarwal-navin) [inline packages/utils/telemetry-utils/src/mockLogger.ts:84]: What does this tag do?

File: packages/runtime/container-runtime/src/test/garbageCollection.spec.ts
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:52]: You can export this from `garbageCollection.ts` and not from `index.ts`. That way it can be used in the `containr-runtime` folder without this being publicly exported from the package.

File: packages/runtime/container-runtime/src/test/garbageCollection.spec.ts
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:476]: What does this mean?

File: packages/runtime/container-runtime/src/test/garbageCollection.spec.ts
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:563]: Nice!

File: packages/runtime/container-runtime/src/test/garbageCollection.spec.ts
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:869]: You can pass this as a param to `createGarbageCollctor` and avoid this hack.

File: packages/runtime/container-runtime/src/test/garbageCollection.spec.ts
Comment (tyler-cai-microsoft) [reply packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:476]: I think it might be better to draw the lines back to the 0 node from the 1 node and the 3 node.

File: packages/runtime/container-runtime/src/test/garbageCollection.spec.ts
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:632]: Why did we disable max-len?

File: packages/runtime/container-runtime/src/test/garbageCollection.spec.ts
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:902]: I wish there was a way to avoid this cast to any :-(

File: packages/runtime/container-runtime/src/garbageCollection.ts
Comment (markfields) [reply packages/runtime/container-runtime/src/garbageCollection.ts:93]: I could also just have independent boolean flags for main v. summarizer.

File: packages/runtime/container-runtime/src/test/garbageCollection.spec.ts
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:902]: Could we call `setConnectionState` that my recent PR added instead? That is a sync function though, so I am not sure how to wait for the initialization to complete.

**Review by agarwal-navin — APPROVED** [review]
Added some comments. Looks good overall.

File: packages/runtime/container-runtime/src/garbageCollection.ts
Comment (markfields) [reply packages/runtime/container-runtime/src/garbageCollection.ts:341]: Yeah, I see that.  How about `unreferencedObjectUsedAfterGarbageCollected`?  I'm dropping the term "Error" at the end because it's so long and don't want it to sound like "after an error with GC"

**Comment by agarwal-navin** [comment]
`Rename all references to "Access Violation"` is probably not applicable now in the description.

File: packages/runtime/container-runtime/src/garbageCollection.ts
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/garbageCollection.ts:93]: I think one flag is better with different values. If you use objects you can run into issues in ECS. You will have to JSON.stringify them in ECS and then parse it here last time I checked.

File: packages/runtime/container-runtime/src/garbageCollection.ts
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/garbageCollection.ts:341]: sounds good.

File: packages/runtime/container-runtime/src/garbageCollection.ts
Comment (markfields) [reply packages/runtime/container-runtime/src/garbageCollection.ts:93]: Ok.  Then I'll stick with the plan I said above.  See https://github.com/markfields/FluidFramework/blob/5af5b089a87711b06a70934ff78e82abd182d28b/packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts#L24-L35

Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/garbageCollection.ts:93]: Oh this is what you meant. I thought you meant adding an object directly in ECS :-).
With the code you shared, either one of these will be set right? How would you set both - summarizerAndmainContainer?

File: packages/runtime/container-runtime/src/test/garbageCollection.spec.ts
Comment (markfields) [reply packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:632]: It's my personal preference - I do it in all test files I work in.  Often test code is more verbose - especially with assert message strings - and I like just letting the lines be long.

I take it you don't like long lines?

File: packages/runtime/container-runtime/src/test/garbageCollection.spec.ts
Comment (markfields) [reply packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:563]: Thanks. After the fact wasn't sure it was worth it, so glad you like it :)  Also it would be cool if mockLogger was more expressive and you could tell it about events you don't want as well.

File: packages/runtime/container-runtime/src/test/garbageCollection.spec.ts
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:632]: Haha yes. It makes it harder to read the code on laptops. Also, the repo rule is to not have lines more than 80 characters :-). Anyway, I was just curious, not asking you to change it.

File: packages/runtime/container-runtime/src/test/garbageCollection.spec.ts
Comment (markfields) [reply packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:476]: Haha it was just to help me, if it's so unclear I'll remove it.

Also, we should rename the string values to 0-3 not 1-4, it's super confusing when debugging a test :P

Comment (markfields) [reply packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:476]: Ok I tried Tyler's idea.  Someone can delete it in a future PR if you don't like it :)

File: packages/runtime/container-runtime/src/test/garbageCollection.spec.ts
Comment (markfields) [reply packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:52]: I wanted it hidden from the code even in garbageCollection.ts since it has special semantics.  I don't mind duplicating magic strings into test files because the test fails if they don't match.

File: packages/runtime/container-runtime/src/test/garbageCollection.spec.ts
Comment (markfields) [reply packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:902]: Eh I'm a lot less worried about `any` in unit tests because if you get it wrong the test will fail.  In production code you can get it wrong but not know because you won't be there when the user runs the code and it's broken.

File: packages/utils/telemetry-utils/src/mockLogger.ts
Comment (markfields) [reply packages/utils/telemetry-utils/src/mockLogger.ts:84]: This is a multi-line string.  Not sure what you mean by tag.

File: packages/runtime/container-runtime/src/test/garbageCollection.spec.ts
Comment (tyler-cai-microsoft) [reply packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:632]: I thought the repo rule was 120?

Comment (markfields) [reply packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:632]: My biggest beef is how it slows down the inner dev loop iterating on tests.  Maybe I will try to adjust my habit to disabling in the file only temporarily but not checking it in (cleaning up before merging).  Feel free to remove and fix up violations sometime.

File: packages/runtime/container-runtime/src/test/garbageCollection.spec.ts
Comment (markfields) [reply packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:902]: (and no `setConnectionState` won't help since it's not async, as you say)

**Pull request merged**
